active/001-tabular-msa.md

```diff
@@ -0,0 +1,335 @@
+no validation and a (slightly) crippled API.
+
+In order to better reflect the non-biological nature of `Sequence` we also
+recommend changing the `repr` to be different from `IUPACSequence`.
+
+Adding `base_frequencies()` as a method to `Sequence` will improve the value of
+returning `Sequence` object for the column.
+
+# Drawbacks
+
+- scikit-bio users will have to update their code to use a new API.
+
+- The new class name `TabularMSA` may not be as obvious as to its behavior as
+the current class name `Alignment`.
+
+- It is unclear if it is advisable to use a tabular (row-column) MSA as it
+forces an ordering where none exists. See
+[Lee et al. 2002](http://bioinformatics.oxfordjournals.org/content/18/3/452.short).
+
+# Alternatives
+
+We considered representing an MSA internally as a DAG in support of
+partial-order MSAs. However, representing positions in such an MSA defeats the
+entire purpose of using a DAG at all.
+
+We considered *only* supporting partial-order MSAs. However, since scikit-bio
+is a bioinformatics data munging library, we need to support row-column MSAs
+because most bioinformatics tools use or produce these structures.
+
+We considered the following alternative names for the new class:
+
+- `LinearMSA`: it is not immediately obvious that this is a row-column
+alignment
+- `TotalOrderMSA`: even less obvious and it sounds "better" than partial-order
+if users aren't clear on the definitions
+- `RowColMSA` / `RowColumnMSA`: painful to type and hard to remember (was it
+"col" or "column"?)
+- `StandardAlignment`: implies "best" or preferred alignment structure
+- `Alignment`: too general as it doesn't describe an MSA, also implies a
+default alignment structure
+- `MatrixMSA`: the name is ambiguous and could be confused with other alignment
+structures such as an alignment scoring matrix. It could also be confused with
+MSAs represented by graphs because graphs can be stored as matrices (e.g.,
+adjacency, incidence). Does not indicate a labeled structure. May imply too
+much about the internal implementation.
+
+# Unresolved questions
+
+What does the interface of `merge` look like? Can both axes be merged?
+If so, do we support the standard SQL join operations? Should it be called
+`join` instead of `merge`?
+
+How should non-standard characters be handled when computing entropy? Currently
+`Alignment's` method `position_entropies` has a keyword
+`nan_on_non_standard_chars`. This seems like a sensible option, however with
+our new boolean vector filtering model, users may be unaware that `nan` will
+always return `False` for any boolean operation. This may result in
+unanticipated behavior depending on how the boolean array was constructed. For
+example, `(ents <= 0.8) & (ents>= 0.3)` is semantically identical to
+`~((ents > 0.8) | (ents < 0.3))` but will handle `nan` in opposite ways. Is it
+acceptable to keep the current behavior or is there a better way to handle
+non-standard characters when computing entropy?
```

Comment:
Hello Gurus,
Jai asked me to weigh in on this and so here it is. The current treatment of non standard characters and gaps in conjunction with entropy filtering is a bit unhelpful and probably incorrect. We have been working on code to treat these in a different way. First, in skbio currently defaults to a base of the natural log = 2.7ish which is just weird. It should probably be a default of 4 w/o gaps and 5 with gaps. The base is the normalization factor and should reflect the total number of characters involved in the entropy calculation. Currently we can Nan anything that isn't a standard character or we can add it into the entropy calculation. I think the best way to handle non degenerate characters is to split them like:    

test_dict['A'] = test_dict['A'] + test_dict['D']/3 + test_dict['H']/3 + test_dict['M']/2 + test_dict['N']/4 + test_dict['R']/2 + test_dict['V']/3 + test_dict['W']/2 

So now the total number of values still adds to the number of sequences (with gaps too) and you don't have to think about degenerate characters any further. You can add an option to turn this off, but I would not as this is probably not the best way to deal with these characters. With the current system, returning Nan just keeps the entropy and doesn't calculate it or filter that position. If nan_on_nonstandard_characters=False then it is calculating entropy on an incorrect base (because they are not added up so it uses the default) and you are not getting the correct entropy values.

Ultimately, I believe gap and degenerate characters should be treated separately and Nan should not be used for gaps. It should be able to toggle between ignoring the gap character and calculating entropy on nucleotides only or include the gap as a character and increase the base by one.